src/isolation/landlock.rs

```diff
@@ -0,0 +1,260 @@
+			compat_level: Self::determine_compat_level(sandbox_mode),
+		}
+	}
+
+	/// Enforce Landlock rulesets.
+	///
+	/// By running [`Self::enforce_landlock`] (and panicking if an error occurs),
+	/// this function is the "public interface" to be used by
+	/// [`UhyveVm::new`](crate::vm::UhyveVm::new) after the necessary rulesets have
+	/// been provided and "stored" in UhyveLandlockWrapper. The restrictions have
+	/// a process-wide effect.
+	pub fn apply_landlock_restrictions(&self) {
+		debug!(
+			"Enabling Landlock path isolation with following compatibility mode: {:?}",
+			self.compat_level
+		);
+
+		let is_already_enabled = LANDLOCK_ENABLED.lock();
+		if *is_already_enabled.unwrap() {
+			warn!(
+				"Landlock has been enabled already. Further policies will not affect existing ones."
+			);
+		}
+
+		{
+			let _status = match Self::enforce_landlock(self) {
+				Ok(status) => status,
+				Err(error) => panic!("Unable to initialize Landlock: {error:?}"),
+			};
+		}
+	}
+
+	/// Using the file sandbox mode provided by the user, we derive the corresponding
+	/// Landlock compatibility mode. This deliberately omits the CompatLevel::SoftRequirement
+	/// possibility, while leaving room for future enhancements to the host file access
+	/// layer.
+	///
+	/// `sandbox_mode` - User-provided [`crate::params::FileSandboxMode`]
+	fn determine_compat_level(sandbox_mode: FileSandboxMode) -> CompatLevel {
+		match sandbox_mode {
+			FileSandboxMode::None => panic!(),
+			FileSandboxMode::Normal => CompatLevel::BestEffort,
+			FileSandboxMode::Strict => CompatLevel::HardRequirement,
+		}
+	}
+
+	/// Internal Landlock enforcement function used by [`Self::apply_landlock_restrictions`].
+	///
+	/// This function directly interfaces with the [`landlock`] crate, and is responsible for
+	/// enforcing the policies defined in [`UhyveVm::new`](crate::vm::UhyveVm::new).
+	fn enforce_landlock(&self) -> Result<RestrictionStatus, RestrictError> {
+		// This should be incremented regularly.
+		let abi = ABI::V5;
+		// Used for explicitly whitelisted files (read & write).
+		let access_all: landlock::BitFlags<AccessFs, u64> = AccessFs::from_all(abi);
+		// Used for the kernel itself, as well as "system directories" that we only read from.
+		let access_read: landlock::BitFlags<AccessFs, u64> = AccessFs::from_read(abi);
+
+		// LANDLOCK_ACCESS_FS_REMOVE_FILE and LANDLOCK_ACCESS_FS_REMOVE_DIR do not apply to
+		// whitelisted files themselves, but apply to the contents of directories transitively.
+		// This means that we have to give Uhyve read-only access to the parent directory with
+		// some additional permissions so as to support fs::remove_file and removing the temporary
+		// directory.
+		let access_dir_with_rm: landlock::BitFlags<AccessFs, u64> =
+			AccessFs::ReadDir | AccessFs::RemoveDir | AccessFs::RemoveFile;
+
+		let res_status = Ruleset::default()
+			.handle_access(access_all)?
+			.create()?
+			.add_rules(
+				self.rw_paths
```

Comment:
TODO: The "RW"/"RO" path distinction breaks things in strict mode, because, if the read dir policy cannot be applied to a file, Landlock enforcement fails.

Reply:
Why doesn't this work, given: https://github.com/landlock-lsm/rust-landlock/blob/a751123a4fc7c3bf69147a2c27699072a52321c4/src/fs.rs#L205-L206

Reply:
See: https://github.com/landlock-lsm/rust-landlock/blob/a751123a4fc7c3bf69147a2c27699072a52321c4/src/fs.rs#L145-L149